modules/parachains/src/lib.rs

```diff
@@ -0,0 +1,594 @@
+		///
+		/// The GRANDPA pallet instance must be configured to import headers of relay chain that
+		/// we're interested in.
+		type BridgesGrandpaPalletInstance: 'static;
+
+		/// Maximal number of single parachain heads to keep in the storage.
+		///
+		/// The setting is there to prevent growing the on-chain state indefinitely. Note
+		/// the setting does not relate to parachain block numbers - we will simply keep as much
+		/// items in the storage, so it doesn't guarantee any fixed timeframe for heads.
+		#[pallet::constant]
+		type HeadsToKeep: Get<u32>;
+	}
+
+	/// Best parachain heads.
+	#[pallet::storage]
+	pub type BestParaHeads<T: Config<I>, I: 'static = ()> =
+		StorageMap<_, Identity, ParaId, BestParaHead>;
+
+	/// Parachain heads which have been imported into the pallet.
+	#[pallet::storage]
+	pub type ImportedParaHeads<T: Config<I>, I: 'static = ()> =
+		StorageDoubleMap<_, Identity, ParaId, Identity, ParaHash, ParaHead>;
+
+	/// A ring buffer of imported parachain head hashes. Ordered by the insertion time.
+	#[pallet::storage]
+	pub(super) type ImportedParaHashes<T: Config<I>, I: 'static = ()> =
+		StorageDoubleMap<_, Identity, ParaId, Identity, u32, ParaHash>;
+
+	#[pallet::pallet]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		<T as pallet_bridge_grandpa::Config<T::BridgesGrandpaPalletInstance>>::BridgedChain:
+			bp_runtime::Chain<
+				BlockNumber = RelayBlockNumber,
+				Hash = RelayBlockHash,
+				Hasher = RelayBlockHasher,
+			>,
+	{
+		/// Submit proof of one or several parachain heads.
+		///
+		/// The proof is supposed to be proof of some `Heads` entries from the
+		/// `polkadot-runtime-parachains::paras` pallet instance, deployed at the bridged chain.
+		/// The proof is supposed to be crafted at the `relay_header_hash` that must already be
+		/// imported by corresponding GRANDPA pallet at this chain.
+		#[pallet::weight(0)] // TODO
+		pub fn submit_parachain_heads(
+			_origin: OriginFor<T>,
+			relay_block_hash: RelayBlockHash,
+			parachains: Vec<ParaId>,
+			parachain_heads_proof: ParachainHeadsProof,
+		) -> DispatchResult {
+			// we'll need relay chain header to verify that parachains heads are always increasing.
+			let relay_block = pallet_bridge_grandpa::ImportedHeaders::<
+				T,
+				T::BridgesGrandpaPalletInstance,
+			>::get(relay_block_hash)
+			.ok_or(Error::<T, I>::UnknownRelayChainBlock)?;
+			let relay_block_number = *relay_block.number();
+
+			// now parse storage proof and read parachain heads
+			pallet_bridge_grandpa::Pallet::<T, T::BridgesGrandpaPalletInstance>::parse_finalized_storage_proof(
+				relay_block_hash,
+				sp_trie::StorageProof::new(parachain_heads_proof),
+				move |storage| {
+					for parachain in parachains {
+						let parachain_head = match Pallet::<T, I>::read_parachain_head(&storage, parachain) {
+							Some(parachain_head) => parachain_head,
+							None => {
```

Comment:
Actually it should be supported - parachain may be offboarded and its head may be removed. So there should be 3 possible outcomes from `read_parachain_head`: `Err(_)`, `Ok(Some(_))` and `Ok(None)`. Right now the first (which is relayer error) is replaced with `Ok(None)`, which is handled incorrectly.

Reply:
I'd prefer to deal with it in follow-up PRs, as it requires some more thinking + there are already two PRs on top of that one.